upstream/upaws/aws.go

```diff
@@ -0,0 +1,98 @@
+package upaws
+
+import (
+	"os"
+	"text/template"
+)
+
+const (
+	AccessEnvVar          = "MM_APPS_AWS_ACCESS_KEY"           // nolint:gosec
+	SecretEnvVar          = "MM_APPS_AWS_SECRET_KEY"           // nolint:gosec
+	ProvisionAccessEnvVar = "MM_APPS_PROVISION_AWS_ACCESS_KEY" // nolint:gosec
+	ProvisionSecretEnvVar = "MM_APPS_PROVISION_AWS_SECRET_KEY" // nolint:gosec
+
+	DeprecatedCloudAccessEnvVar = "APPS_INVOKE_AWS_ACCESS_KEY" // nolint:gosec
+	DeprecatedCloudSecretEnvVar = "APPS_INVOKE_AWS_SECRET_KEY" // nolint:gosec
+
+	// S3BucketEnvVar is the environment variable containing the S3 bucket name
+	// used to host Apps' assets.
+	S3BucketEnvVar  = "MM_APPS_S3_BUCKET"
+	DefaultS3Bucket = "mattermost-apps-bucket"
+
+	RegionEnvVar  = "MM_APPS_AWS_REGION"
+	DefaultRegion = "us-east-1"
+)
+
+const LambdaExecutionPolicyARN = ARN(`arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole`)
+
+const (
+	DefaultExecuteRoleName = "mattermost-apps-execute-lambda-role"
+	DefaultPolicyName      = "mattermost-apps-invoke-policy"
+	DefaultUserName        = "mattermost-apps-invoke"
+	DefaultGroupName       = "mattermost-apps-invoke-group"
+)
+
+const AssumeRolePolicyDocument = `{
+	"Version": "2012-10-17",
+	"Statement": [
+		{
+			"Effect": "Allow",
+			"Principal": {
+				"Service": "lambda.amazonaws.com"
+			},
+			"Action": "sts:AssumeRole"
+		}
+	]
+}`
+
+const InitialInvokePolicyDocument = `{
```

Comment:
@levb Have you tested this? You may need to add this section too
```json
{
            "Effect": "Allow",
            "Action": [
                "s3:ListBucket",
                "s3:ListBucketVersions",
                "s3:GetBucketLocation"
            ],
            "Resource": [
                ""arn:aws:s3:::{{.Bucket}}"
            ]
        }
```

Reply:
Yes, we did (with **s, still a final round to be done). From https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html

You need the s3:GetObject permission for this operation. For more information, see Specifying Permissions in a Policy. If the object you request does not exist, the error Amazon S3 returns depends on whether you also have the s3:ListBucket permission.

If you have the s3:ListBucket permission on the bucket, Amazon S3 will return an HTTP status code 404 ("no such key") error.

If you don’t have the s3:ListBucket permission, Amazon S3 will return an HTTP status code 403 ("access denied") error.

Reply:
Getting a 404 is fine